### PR TITLE
udp: add source-specific multicast support

### PR DESCRIPTION
**Pull request by falanger**
Add [RFC 4607](https://tools.ietf.org/html/rfc4607) support for IPv4 and IPv6.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -74,5 +74,7 @@ ipch
 *.xcodeproj
 *.xcworkspace
 
+CMakeLists.txt
```
Comment (saghul) [inline .gitignore]: unneeded change, please drop it

File: docs/src/udp.rst [hunk 2]
```diff
@@ -181,6 +181,23 @@ API
 
     :returns: 0 on success, or an error code < 0 on failure.
 
+.. c:function:: int uv_udp_set_source_membership(uv_udp_t* handle, const char* multicast_addr, const char* interface_addr, const char* source_addr, uv_membership membership)
+
+    Set membership for a source-specific multicast group. Currently implemented only for IPv4.
```
Comment (saghul) [inline docs/src/udp.rst]: What's the story for IPv6? I tried to look it up, but other than some Cisco IOS docs and some FreeBSD stuff there is not much I can find...

File: include/uv.h [hunk 3]
```diff
@@ -617,6 +617,11 @@ UV_EXTERN int uv_udp_set_membership(uv_udp_t* handle,
                                     const char* multicast_addr,
                                     const char* interface_addr,
                                     uv_membership membership);
+UV_EXTERN int uv_udp_set_source_membership(uv_udp_t* handle,
+                                    const char* multicast_addr,
```
Comment (saghul) [inline include/uv.h]: style: please align the fields

File: docs/src/udp.rst [hunk 4]
```diff
@@ -181,6 +181,23 @@ API
...
+
+    :param multicast_addr: Multicast address to set membership for.
+
+    :param interface_addr: Interface address.
+
+    :param source_addr: Source address.
+
+    :param membership: Should be ``UV_JOIN_GROUP`` or ``UV_LEAVE_GROUP``.
+
+    :returns: 0 on success, or an error code < 0 on failure.
```
Comment (saghul) [inline docs/src/udp.rst]: You'll need to add a `.. versionadded:: 1.10.0` here

File: src/unix/udp.c [hunk 5]
```diff
@@ -557,6 +557,51 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
 }
 
 
+static int uv__udp_set_source_membership4(uv_udp_t* handle,
+                                   const struct sockaddr_in* multicast_addr,
+                                   const char* interface_addr,
+                                   const struct sockaddr_in* source_addr,
+                                   uv_membership membership) {
```
Comment (saghul) [inline src/unix/udp.c]: style: align

File: src/unix/udp.c [hunk 6]
```diff
@@ -639,6 +684,29 @@ int uv_udp_set_membership(uv_udp_t* handle,
   }
 }
 
+
+int uv_udp_set_source_membership(uv_udp_t* handle,
+                          const char* multicast_addr,
```
Comment (saghul) [inline src/unix/udp.c]: style: align

File: src/unix/udp.c [hunk 7]
```diff
@@ -639,6 +684,29 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+                          const char* interface_addr,
+                          const char* source_addr,
+                          uv_membership membership) {
+  struct sockaddr_in mcast_addr4;
+  struct sockaddr_in src_addr4;
+
+  int err = uv_ip4_addr(multicast_addr, 0, &mcast_addr4);
+  if (!err) {
```
Comment (saghul) [inline src/unix/udp.c]: can you reverse the logic here? that way we save one level of indentation

File: src/win/udp.c [hunk 8]
```diff
@@ -700,6 +700,62 @@ int uv__udp_set_membership6(uv_udp_t* handle,
 }
 
 
+static int uv__udp_set_source_membership4(uv_udp_t* handle,
+                                   const struct sockaddr_in* multicast_addr,
```
Comment (saghul) [inline src/win/udp.c]: style: align

File: src/win/udp.c [hunk 9]
```diff
@@ -716,6 +772,25 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+int uv_udp_set_source_membership(uv_udp_t* handle,
+                          const char* multicast_addr,
+                          const char* interface_addr,
+                          const char* source_addr,
+                          uv_membership membership) {
+  struct sockaddr_in mcast_addr4;
+  struct sockaddr_in src_addr4;
+
+  int err = uv_ip4_addr(multicast_addr, 0, &mcast_addr4);
+  if (!err) {
```
Comment (saghul) [inline src/win/udp.c]: same as above, reverse the logic

File: src/unix/udp.c [hunk 10]
```diff
@@ -639,6 +684,29 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+    err = uv__udp_maybe_deferred_bind(handle, AF_INET, UV_UDP_REUSEADDR);
```
Comment (saghul) [inline src/unix/udp.c]: maybe move this to inside uv__udp_set_source_membership4 like on Windows

**Comment by saghul** [comment]
Thanks for contributing @falanger! Mostly LGTM, left some comments.

File: docs/src/udp.rst (hunk 2 above)
Comment (falanger) [inline docs/src/udp.rst]: The modern network API's don't provide IPv6-specific `mreq_source` structure. Recommended way to implement SSM in IPv6 is [group_source_req](https://msdn.microsoft.com/en-us/library/windows/desktop/bb427441(v=vs.85).aspx) usage, but this may broke back-compatibility in libuv.

**Comment by falanger** [comment]
Hi, @saghul! Thanks for your time! I've fixed all the comments.

File: docs/src/udp.rst (hunk 2 above)
Comment (saghul) [inline docs/src/udp.rst]: Can you elaborate? How would that brek backwards compatibility?

**Comment by saghul** [comment]
Changes look good, though I'd like to add IPv6 support if possible. /cc @libuv/collaborators any other reviews?

File: docs/src/udp.rst (hunk 2 above)
Comment (santigimeno) [inline docs/src/udp.rst]: There's a good example from `Stevens Unix Network Programming` that supports IPv4 and IPv6 [here](https://github.com/harrifeng/unpbook/blob/master/lib/mcast_join.c#L105-L173) that might help. As @falanger comments it uses `MCAST_JOIN_SOURCE_GROUP` option with the `group_source_req` structure for `IPv6`.

File: src/unix/udp.c [hunk 11]
```diff
@@ -557,6 +557,55 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+
+  switch (membership) {
+  case UV_JOIN_GROUP:
+    optname = IP_ADD_SOURCE_MEMBERSHIP;
+    break;
+  case UV_LEAVE_GROUP:
+    optname = IP_DROP_SOURCE_MEMBERSHIP;
+    break;
+  default:
+    return -EINVAL;
```
Comment (santigimeno) [inline src/unix/udp.c]: I think one level of indentation is missing here.

Comment (falanger) [inline src/unix/udp.c]: It's consistent with other "switch..case" instructions in this file.

**Comment by falanger** [comment]
@saghul sorry, I was wrong about IPv6 support (thanks, @santigimeno).

**Comment by falanger** [comment]
@saghul @santigimeno any comments?

File: src/unix/udp.c [hunk 12]
```diff
@@ -639,6 +753,34 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+  struct sockaddr_in src_addr4;
+  struct sockaddr_in6 mcast_addr6;
+  struct sockaddr_in6 src_addr6;
+
+  int err = uv_ip4_addr(multicast_addr, 0, &mcast_addr4);
+  if (err) {
+    err = uv_ip6_addr(multicast_addr, 0, &mcast_addr6);
+    if (err)
+        return err;
+    err = uv_ip6_addr(multicast_addr, 0, &src_addr6);
```
Comment (santigimeno) [inline src/unix/udp.c]: shouldn't this be `source_addr`?

File: src/unix/udp.c (hunk 11 above)
Comment (santigimeno) [inline src/unix/udp.c]: Even though the style is not consistent in the project, I would add indentation here as it's already the most commonly used format in the source and I think consistency is good. Anyway, let's see what others think.

File: src/unix/udp.c [hunk 13]
```diff
@@ -557,6 +568,109 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+
+  memset(&mreq, 0, sizeof mreq);
+
+  if (interface_addr) {
+    if (uv_ip6_addr(interface_addr, 0, &addr6))
+      return -EINVAL;
+    mreq.gsr_interface = addr6.sin6_scope_id;
+  } else {
+    mreq.gsr_interface = 0;
+  }
```
Comment (santigimeno) [inline src/unix/udp.c]: isn't `mreq.gsr_interface` already 0?

File: src/unix/udp.c [hunk 14]
```diff
@@ -557,6 +568,109 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+
+  err = uv__udp_maybe_deferred_bind(handle, AF_INET6, UV_UDP_REUSEADDR);
+  if (err)
+    return err;
+
+  memset(&mreq, 0, sizeof mreq);
+
+  if (interface_addr) {
+    if (uv_ip6_addr(interface_addr, 0, &addr6))
+      return -EINVAL;
```
Comment (santigimeno) [inline src/unix/udp.c]: why don't return what `uv_ip6_addr` returns instead of `-EINVAL`?

File: src/win/udp.c [hunk 15]
```diff
@@ -700,6 +711,123 @@ int uv__udp_set_membership6(uv_udp_t* handle,
...
+  switch (membership) {
+  case UV_JOIN_GROUP:
+    optname = IPV6_ADD_SOURCE_MEMBERSHIP;
+    break;
+  case UV_LEAVE_GROUP:
+    optname = IPV6_DROP_SOURCE_MEMBERSHIP;
+    break;
+  default:
+    return UV_EINVAL;
+  }
```
Comment (santigimeno) [inline src/win/udp.c]: Indentation as in the previous switch in the file

File: src/unix/udp.c [hunk 16]
```diff
@@ -557,6 +568,109 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+                                          uv_membership membership) {
+  struct ip_mreq_source mreq;
+  int optname;
+  int err;
+
+  err = uv__udp_maybe_deferred_bind(handle, AF_INET, UV_UDP_REUSEADDR);
+  if (err)
+    return err;
+
+  memset(&mreq, 0, sizeof mreq);
```
Comment (cjihrig) [inline src/unix/udp.c]: Could you use consistent parentheses with `sizeof`.

**Comment by LPardue** [comment]
I was hoping to do some source-specific multicast in node.js (which uses libuv). This PR would introduce that capability but it is unfortunately stalled, is there any way to unblock it? It seemed like the last review identified some minor style issues but otherwise things were ok.

File: src/win/udp.c [hunk 17]
```diff
@@ -702,6 +713,123 @@ int uv__udp_set_membership6(uv_udp_t* handle,
...
+  if (setsockopt(handle->socket,
+                 IPPROTO_IPV6,
+                 optname,
+                 (char*) &mreq,
+                 sizeof mreq) == SOCKET_ERROR) {
+    return uv_translate_sys_error(WSAGetLastError());
+  }
+
+  return 0;
+#else
```
Comment (cjihrig) [inline src/win/udp.c]: Can you add a comment here that mentions what `#ifdef` this corresponds to.

File: docs/src/udp.rst [hunk 18]
```diff
@@ -181,6 +181,25 @@ API
 
     :returns: 0 on success, or an error code < 0 on failure.
 
+.. c:function:: int uv_udp_set_source_membership(uv_udp_t* handle, const char* multicast_addr, const char* interface_addr, const char* source_addr, uv_membership membership)
+
+    Set membership for a source-specific multicast group
```
Comment (cjihrig) [inline docs/src/udp.rst]: Missing period at the end of the line.

File: docs/src/udp.rst [hunk 19]
```diff
@@ -181,6 +181,25 @@ API
...
+
+    :param interface_addr: Interface address.
+
+    :param source_addr: Source address.
+
+    :param membership: Should be ``UV_JOIN_GROUP`` or ``UV_LEAVE_GROUP``.
+
+    :returns: 0 on success, or an error code < 0 on failure.
+
+    .. versionadded:: 1.10.0
```
Comment (cjihrig) [inline docs/src/udp.rst]: This version number needs to be updated. Currently, this would be 1.15.0 at the earliest.

**Comment by cjihrig** [comment]
@falanger it's been quite a while. Sorry about that. Are you still interested in pursuing this PR?

**Comment by falanger** [comment]
@cjihrig Hi! Yes, I am interested in. I'll try to fix issues on this week.

File: src/unix/udp.c (hunk 12 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: src/unix/udp.c (hunk 13 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: src/unix/udp.c (hunk 14 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: src/unix/udp.c (hunk 16 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: docs/src/udp.rst (hunk 18 above)
Comment (falanger) [reply docs/src/udp.rst]: fixed

File: docs/src/udp.rst (hunk 19 above)
Comment (falanger) [reply docs/src/udp.rst]: fixed

File: src/win/udp.c (hunk 17 above)
Comment (falanger) [reply src/win/udp.c]: fixed

File: test/test-udp-multicast-join6.c [hunk 20]
```diff
@@ -134,6 +134,17 @@ TEST_IMPL(udp_multicast_join6) {
 
   ASSERT(r == 0);
 
+  r = uv_udp_set_source_membership(&client, "ff02::1", NULL, "2001:db8:a0b:12f0::1", UV_JOIN_GROUP);
```
Comment (cjihrig) [inline test/test-udp-multicast-join6.c]: Can the rest of the test still be run if `UV_ENODEV` or `UV_EPROTONOSUPPORT` is returned? If so, maybe the assertion should just allow them so that the remainder of the test is still executed. If that can't be done, maybe a separate test file should be created.

File: test/test-udp-multicast-join.c [hunk 21]
```diff
@@ -120,6 +120,12 @@ TEST_IMPL(udp_multicast_join) {
     RETURN_SKIP("No multicast support.");
   ASSERT(r == 0);
 
+  /* join the source-specific multicast channel */
+  r = uv_udp_set_source_membership(&client, "239.255.0.1", NULL, "10.50.129.200", UV_JOIN_GROUP);
```
Comment (cjihrig) [inline test/test-udp-multicast-join.c]: Same comment here about finishing the rest of the test.

File: src/win/udp.c [hunk 22]
```diff
@@ -716,6 +843,33 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+  struct sockaddr_in mcast_addr4;
+  struct sockaddr_in src_addr4;
+  struct sockaddr_in6 mcast_addr6;
+  struct sockaddr_in6 src_addr6;
+
+  int err = uv_ip4_addr(multicast_addr, 0, &mcast_addr4);
+  if (err) {
+    err = uv_ip6_addr(multicast_addr, 0, &mcast_addr6);
+    if (err)
+        return err;
```
Comment (cjihrig) [inline src/win/udp.c]: Two space indentation please.

File: src/win/udp.c [hunk 23]
```diff
@@ -716,6 +843,33 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+    err = uv_ip6_addr(source_addr, 0, &src_addr6);
+    if (err)
+        return err;
```
Comment (cjihrig) [inline src/win/udp.c]: And here.

File: src/win/udp.c [hunk 24]
```diff
@@ -692,11 +703,127 @@ int uv__udp_set_membership6(uv_udp_t* handle,
...
+                 IPPROTO_IPV6,
+                 optname,
+                 (char*) &mreq,
+                 sizeof(mreq)) == SOCKET_ERROR) {
     return uv_translate_sys_error(WSAGetLastError());
   }
 
   return 0;
+#else
+  return -EPROTONOSUPPORT;
```
Comment (cjihrig) [inline src/win/udp.c]: `UV_EPROTONOSUPPORT`

File: src/unix/udp.c [hunk 25]
```diff
@@ -658,6 +771,34 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+  struct sockaddr_in mcast_addr4;
+  struct sockaddr_in src_addr4;
+  struct sockaddr_in6 mcast_addr6;
+  struct sockaddr_in6 src_addr6;
+
+  int err = uv_ip4_addr(multicast_addr, 0, &mcast_addr4);
+  if (err) {
+    err = uv_ip6_addr(multicast_addr, 0, &mcast_addr6);
+    if (err)
+        return err;
```
Comment (cjihrig) [inline src/unix/udp.c]: Again, two space indentation. I'll stop commenting on it :-)

File: src/unix/udp.c [hunk 26]
```diff
@@ -490,7 +501,7 @@ static int uv__udp_set_membership4(uv_udp_t* handle,
   int optname;
   int err;
 
-  memset(&mreq, 0, sizeof mreq);
+  memset(&mreq, 0, sizeof(mreq));
```
Comment (cjihrig) [inline src/unix/udp.c]: Can you revert all of the unrelated style changes.

File: src/win/udp.c (hunk 24 above)
Comment (falanger) [reply src/win/udp.c]: fixed

File: src/win/udp.c (hunk 22 above)
Comment (falanger) [reply src/win/udp.c]: Sorry, I have fixed all indentation issues.

File: src/unix/udp.c (hunk 26 above)
Comment (falanger) [reply src/unix/udp.c]: Fixed. BTW, why the style is inconsistent in file (and project) scope?

Comment (cjihrig) [reply src/unix/udp.c]: > why the style is inconsistent in file (and project) scope?

Because we haven't agreed on and enforced automated linting. So, some stuff gets by the human eye test.

File: src/unix/udp.c [hunk 27]
```diff
@@ -576,6 +587,108 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+    if (err)
+      return err;
+  } else {
+    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+  }
+
+  mreq.imr_multiaddr.s_addr = multicast_addr->sin_addr.s_addr;
+  mreq.imr_sourceaddr.s_addr = source_addr->sin_addr.s_addr;
+
+  switch (membership) {
```
Comment (cjihrig) [inline src/unix/udp.c]: I would just write this as an `if...else if...else`. It would be several lines shorter, and there are only a couple of options. I'd also move it to the top of the function so that `-EINVAL` can be returned right away if necessary.

File: src/unix/udp.c [hunk 28]
```diff
@@ -576,6 +587,108 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+                 IPPROTO_IPV6,
+                 optname,
+                 &mreq,
+                 sizeof(mreq))) {
+    return -errno;
+  }
+
+  return 0;
+#else
+  return UV_EPROTONOSUPPORT;
```
Comment (cjihrig) [inline src/unix/udp.c]: I think this should be `-EPROTONOSUPPORT`.

File: src/unix/udp.c [hunk 29]
```diff
@@ -576,6 +587,108 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+    err = uv_ip6_addr(interface_addr, 0, &addr6);
+    if (err)
+      return err;
+    mreq.gsr_interface = addr6.sin6_scope_id;
+  }
+
+  memcpy(&mreq.gsr_group, multicast_addr, sizeof(mreq.gsr_group));
+  memcpy(&mreq.gsr_source, source_addr, sizeof(mreq.gsr_source));
+
+  switch (membership) {
```
Comment (cjihrig) [inline src/unix/udp.c]: Same comment as the previous `switch` statement.

File: src/unix/udp.c (hunk 28 above)
Comment (cjihrig) [reply src/unix/udp.c]: Come to think of it, I think we return `ENOSYS` in similar situations, but I could be wrong.

File: src/win/udp.c [hunk 30]
```diff
@@ -28,6 +28,17 @@
 #include "stream-inl.h"
 #include "req-inl.h"
 
+#if defined(MCAST_JOIN_SOURCE_GROUP) && defined(MCAST_LEAVE_SOURCE_GROUP)
```
Comment (cjihrig) [inline src/win/udp.c]: Would it be possible to consolidate with the same code in `src/unix/udp.c`?

File: src/win/udp.c [hunk 31]
```diff
@@ -700,6 +711,122 @@ int uv__udp_set_membership6(uv_udp_t* handle,
...
+    if (err)
+      return err;
+  } else {
+    mreq.imr_interface.s_addr = htonl(INADDR_ANY);
+  }
+
+  mreq.imr_multiaddr.s_addr = multicast_addr->sin_addr.s_addr;
+  mreq.imr_sourceaddr.s_addr = source_addr->sin_addr.s_addr;
+
+  switch (membership) {
```
Comment (cjihrig) [inline src/win/udp.c]: Same comment about the `switch`. I'll stop repeating it now.

File: src/win/udp.c [hunk 32]
```diff
@@ -700,6 +711,122 @@ int uv__udp_set_membership6(uv_udp_t* handle,
...
+                          sizeof(uv_addr_ip6_any_),
+                          UV_UDP_REUSEADDR);
+
+  if (err)
+    return uv_translate_sys_error(err);
+
+  memset(&mreq, 0, sizeof(mreq));
+
+  if (interface_addr) {
+    err = uv_ip6_addr(interface_addr, 0, &addr6)
```
Comment (LPardue) [inline src/win/udp.c]: Missing semi-colon, causes a compile failure

File: src/win/udp.c [hunk 33]
```diff
@@ -700,6 +711,122 @@ int uv__udp_set_membership6(uv_udp_t* handle,
...
+    if (err)
+      return err;
+    mreq.ipv6mr_interface = addr6.sin6_scope_id;
```
Comment (LPardue) [inline src/win/udp.c]: This fails to compile for me:

> src\win\udp.c(798): error C2039: 'ipv6mr_interface': is not a member of 'group_source_req'

Do you mean `mreq.gsr_interface`?

File: src/win/udp.c (hunk 32 above)
Comment (falanger) [reply src/win/udp.c]: fixed

File: src/win/udp.c (hunk 33 above)
Comment (falanger) [reply src/win/udp.c]: Yes, that is. Fixed

Comment (falanger) [reply src/win/udp.c]: @LPardue please, send me a mention next time. I'll try to write tests as soon as I can - it's the last thing that I need to be done to merge this PR.

File: src/unix/udp.c (hunk 27 above)
Comment (falanger) [reply src/unix/udp.c]: I agree with you, but in this case functions `uv_udp_set_membership` and `uv_udp_set_source_membership` would be written in different manner. If this OK, I would replace `switch .. case` with `if .. else` construction.

File: src/unix/udp.c (hunk 28 above)
Comment (falanger) [reply src/unix/udp.c]: In my opinion, `ENOSYS` means that function not implemented, so I think that `EPROTONOSUPPORT` or `ENOPROTOOPT` is more suitable error in this case.

File: src/win/udp.c (hunk 30 above)
Comment (falanger) [reply src/win/udp.c]: Yes, It's possible. Can I place this defines in `uv.h` header?

File: src/win/udp.c (hunk 33 above)
Comment (LPardue) [reply src/win/udp.c]: @falanger will do, feel free to take inspiration from the SSM node tests I created in https://github.com/nodejs/node/pull/15735

File: docs/src/udp.rst [hunk 34]
```diff
@@ -181,6 +181,25 @@ API
...
+
+    :param interface_addr: Interface address.
+
+    :param source_addr: Source address.
+
+    :param membership: Should be ``UV_JOIN_GROUP`` or ``UV_LEAVE_GROUP``.
+
+    :returns: 0 on success, or an error code < 0 on failure.
+
+    .. versionadded:: 1.15.0
```
Comment (cjihrig) [inline docs/src/udp.rst]: This would now be 1.16.0

File: src/unix/udp.c (hunk 27 above)
Comment (cjihrig) [reply src/unix/udp.c]: Yea, I'd go with `if...else`

File: src/unix/udp.c [hunk 35]
```diff
@@ -576,6 +587,108 @@ static int uv__udp_set_membership6(uv_udp_t* handle,
...
+  int optname;
+  int err;
+
+  err = uv__udp_maybe_deferred_bind(handle, AF_INET, UV_UDP_REUSEADDR);
+  if (err)
+    return err;
+
+  memset(&mreq, 0, sizeof(mreq));
+
+  if (interface_addr) {
```
Comment (cjihrig) [inline src/unix/udp.c]: One thing throughout this PR. Can you try to be more explicit in these `if`s by checking for `==` and `!=` to 0, `NULL`, etc.

File: src/unix/udp.c (hunk 28 above)
Comment (cjihrig) [reply src/unix/udp.c]: I'll defer to other @libuv/collaborators. Either way, this should be `-EWHATEVER` in the Unix code.

**Comment by BridgeAR** [comment]
Seems like this is required to unblock a PR in Node.js. It would be great if we could get this to land some time soon therefore :-)

**Comment by BridgeAR** [comment]
Ping @cjihrig @saghul @bnoordhuis

**Comment by cjihrig** [comment]
Unnecessary ping. It looks like there are a few unresolved review comments.

**Comment by BridgeAR** [comment]
@cjihrig you are right. Sorry, I did not check the comments again. I looked at to many PRs today.

@falanger can you please have a look at the comments?

File: src/unix/udp.c (hunk 27 above)
Comment (falanger) [reply src/unix/udp.c]: done

File: src/unix/udp.c (hunk 29 above)
Comment (falanger) [reply src/unix/udp.c]: done

File: src/unix/udp.c (hunk 28 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: src/win/udp.c (hunk 30 above)
Comment (falanger) [reply src/win/udp.c]: moved to uv.h

File: src/win/udp.c (hunk 31 above)
Comment (falanger) [reply src/win/udp.c]: done

File: docs/src/udp.rst (hunk 34 above)
Comment (falanger) [reply docs/src/udp.rst]: fixed

File: src/unix/udp.c (hunk 35 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

**Comment by falanger** [comment]
@cjihrig Hi! I've resolved all issues.

File: docs/src/udp.rst [hunk 36]
```diff
@@ -181,6 +181,25 @@ API
...
+
+    :param interface_addr: Interface address.
+
+    :param source_addr: Source address.
+
+    :param membership: Should be ``UV_JOIN_GROUP`` or ``UV_LEAVE_GROUP``.
+
+    :returns: 0 on success, or an error code < 0 on failure.
+
+    .. versionadded:: 1.16.0
```
Comment (BridgeAR) [inline docs/src/udp.rst]: I guess it should be `Version 1.20.0` now because the current version is already `Version 1.19.1`.

Comment (falanger) [reply docs/src/udp.rst]: fixed

File: src/unix/udp.c [hunk 37]
```diff
@@ -660,6 +754,34 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+
+int uv_udp_set_source_membership(uv_udp_t* handle,
+                                 const char* multicast_addr,
+                                 const char* interface_addr,
+                                 const char* source_addr,
+                                 uv_membership membership) {
+  struct sockaddr_in mcast_addr4;
+  struct sockaddr_in src_addr4;
+  struct sockaddr_in6 mcast_addr6;
+  struct sockaddr_in6 src_addr6;
```
Comment (santigimeno) [inline src/unix/udp.c]: I think you could use 2 `sockaddr_storage` that you could later on cast to `sockaddr_in` pr `sockaddr_in6`. See: `uv_udp_set_multicast_interface` implementation.

File: src/unix/udp.c [hunk 38]
```diff
@@ -660,6 +754,34 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+
+  int err = uv_ip4_addr(multicast_addr, 0, &mcast_addr4);
```
Comment (santigimeno) [inline src/unix/udp.c]: style: we separate variable declaration from assignment.

File: src/win/udp.c [hunk 39]
```diff
@@ -716,6 +824,33 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
 
+int uv_udp_set_source_membership(uv_udp_t* handle,
+                                 const char* multicast_addr,
+                                 const char* interface_addr,
+                                 const char* source_addr,
+                                 uv_membership membership) {
+  struct sockaddr_in mcast_addr4;
+  struct sockaddr_in src_addr4;
+  struct sockaddr_in6 mcast_addr6;
+  struct sockaddr_in6 src_addr6;
```
Comment (santigimeno) [inline src/win/udp.c]: Same comments as on `unix`.

File: src/unix/udp.c (hunk 38 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: src/unix/udp.c (hunk 37 above)
Comment (falanger) [reply src/unix/udp.c]: I don't think so because the `uv_udp_set_source_membership` function should reproduce `uv_udp_set_membership` functionality with additional parameter `source_addr`. These functions have the same behavior and they should be written in the same manner IMHO.

File: src/unix/udp.c [hunk 40]
```diff
@@ -660,6 +754,35 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+      return err;
+    err = uv_ip6_addr(source_addr, 0, &src_addr6);
+    if (err)
+      return err;
+    return uv__udp_set_source_membership6(handle, &mcast_addr6, interface_addr, &src_addr6, membership);
+  }
+  err = uv_ip4_addr(source_addr, 0, &src_addr4);
+  if (err)
+    return err;
+  return uv__udp_set_source_membership4(handle, &mcast_addr4, interface_addr, &src_addr4, membership);
```
Comment (santigimeno) [inline src/unix/udp.c]: Style: long lines. Max should be 80 characters.

File: src/win/udp.c [hunk 41]
```diff
@@ -716,6 +824,34 @@ int uv_udp_set_membership(uv_udp_t* handle,
...
+      return err;
+    err = uv_ip6_addr(source_addr, 0, &src_addr6);
+    if (err)
+      return err;
+    return uv__udp_set_source_membership6(handle, &mcast_addr6, interface_addr, &src_addr6, membership);
+  }
+  err = uv_ip4_addr(source_addr, 0, &src_addr4);
+  if (err)
+    return err;
+  return uv__udp_set_source_membership4(handle, &mcast_addr4, interface_addr, &src_addr4, membership);
```
Comment (santigimeno) [inline src/win/udp.c]: Style: long lines

**Comment by santigimeno** [comment]
Are you planning on adding tests?

File: src/unix/udp.c (hunk 37 above)
Comment (falanger) [reply src/unix/udp.c]: Ok, I'll take a look!

File: src/unix/udp.c (hunk 40 above)
Comment (falanger) [reply src/unix/udp.c]: fixed

File: src/win/udp.c (hunk 41 above)
Comment (falanger) [reply src/win/udp.c]: fixed

**Comment by falanger** [comment]
Sure, I plan to write tests for join/leave SSM group. Unfortunately, it not possible to write tests that receive/send a multicast packet (just like a "usual" multicast) because they require the "real" multicast network.

**Comment by BridgeAR** [comment]
Ping @falanger

**Comment by BridgeAR** [comment]
I see there was a push to the branch after my last comment but I am not sure what the status of this is @falanger

**Comment by BridgeAR** [comment]
Ping @falanger

**Comment by BridgeAR** [comment]
@santigimeno would you be willing to add some tests for this? It would really be great to unblock the Node.js PR by getting this ready and to land it.

**Comment by LPardue** [comment]
A pragmatic decision would be to just have a simple test that shows data flow between sender and receiver in a similar vein to the existing any source multicast. Full testing of source-specific multicast needs an environment more complex than what the framework provides (e.g. multiple network subnets and a router).

It appears this ticket is stalled on that issue and there is lack of input from the project on what is an appropriate solution.

**Comment by OlofBailey** [comment]
What's the status of this lack of SSM support in node.js is turning into a major pain for a lot of projects.

**Comment by santigimeno** [comment]
Superseded by https://github.com/libuv/libuv/pull/2202.
